### ISSUE TITLE
[REVIEW]: cuIBM: A GPU-based immersed boundary method code

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#301]
**Submitting author:** @mesnardo (<a href="http://orcid.org/0000-0001-5335-7853">Olivier Mesnard</a>)
**Repository:** <a href="https://github.com/barbagroup/cuIBM" target ="_blank">https://github.com/barbagroup/cuIBM</a>
**Version:** v0.1
**Editor:** @kyleniemeyer
**Reviewer:** @arghdos
**Archive:** <a href="http://dx.doi.org/10.5281/zenodo.832338" target="_blank">10.5281/zenodo.832338</a>

## Status

[![status](http://joss.theoj.org/papers/885ae94611e88776487cd29eee05fd71/status.svg)](http://joss.theoj.org/papers/885ae94611e88776487cd29eee05fd71)

Status badge code:

```
HTML: <a href="http://joss.theoj.org/papers/885ae94611e88776487cd29eee05fd71"><img src="http://joss.theoj.org/papers/885ae94611e88776487cd29eee05fd71/status.svg"></a>
Markdown: [![status](http://joss.theoj.org/papers/885ae94611e88776487cd29eee05fd71/status.svg)](http://joss.theoj.org/papers/885ae94611e88776487cd29eee05fd71)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/barbagroup/cuIBM" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer questions

@arghdos, please carry out your review in this issue by updating the checklist below (please make sure you're logged in to GitHub). The reviewer guidelines are available here: http://joss.theoj.org/about#reviewer_guidelines. Any questions/concerns please let @kyleniemeyer know.

### Conflict of interest

- [x] As the reviewer I confirm that there are no conflicts of interest for me to review this work (such as being a major contributor to the software).

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](http://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/barbagroup/cuIBM">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** Does the release version given match the GitHub release (v0.1)?
- [x] **Authorship:** Has the submitting author (@mesnardo) made major contributions to the software?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** Have any performance claims of the software been confirmed?

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

**Comment by whedon** [comment]
Hello human, I'm @whedon. I'm here to help you with some common editorial tasks for JOSS. @arghdos it looks like you're currently assigned as the reviewer for this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As as reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all JOSS reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

**Comment by kyleniemeyer** [comment]
@arghdos and @osumageed, this is where the review will take place.

@arghdos since you are the "primary" reviewer (and JOSS is only really set up to handle one reviewer), I'll ask you to check off the boxes at the top. However @osumageed you should also go through and make sure you agree that they are satisfied. Thanks!

**Comment by arghdos** [comment]
@kyleniemeyer @mesnardo

I have completed a preliminary review of cuIBM.  This is a valuable software that enables CUDA based solution of the 2D incompressible Navier-Stokes equations using immersed boundary methods.  Installation is relatively easy, and is laid out comprehensively in the Readme, there are many examples for use, and there is API level documentation of the doe. 

Most of the requirements have been met, but a few issues remain.  Two simple issues have been opened on the cuIBM repo:

1.  [#32](https://github.com/barbagroup/cuIBM/issues/32) - Add a link to the API to the readme for easier access to users.
2.  [#33](https://github.com/barbagroup/cuIBM/issues/33) - Add contribution guidlines, as required for JOSS acceptance (there are examples given in the link left in the issue)

However, there is an open question of on validation of performance.  In one of the early papers linked to in this repo, there are claims of up to 4-8x solution speedup over a single CPU core.  However, it appears the CPU implementation of cuIBM is [not currently functional](https://github.com/barbagroup/cuIBM/issues/20) hence there is no way to validate the performance of cuIBM against any "reference technique".  I am not sure how to proceed with this issue

**Comment by OSUmageed** [comment]
@kyleniemeyer @arghdos @mesnardo

I completely agree with @arghdos' preliminary review.  And I have a few things to add.

First, I want to say the documentation is really spectacular.  I'm still relatively new to this field, but this is the most well documented source code I've ever read.  

I definitely think there needs to be a link to the API in the readme or instructions to build and view it, if only because it allows the user to copy and paste multiple lines of code into a terminal at once which the > symbol in the readme prevents.

The lack of performance validation is an issue.  It doesn't need to be a full example; a study with plots and an adequate description of the hardware that produced them would suffice.  The performance is essential to the **statement of need**.  If it's not faster on the GPU, why not write it in C/C++?

I also opened an issue in the repo [#34](https://github.com/barbagroup/cuIBM/issues/34).  The plotting scripts for the flying snake examples don't work.

**Comment by mesnardo** [comment]
Thank you @arghdos for the preliminary review and for your suggestions.

I have added a link to the Doxygen API documentation in the README of cuIBM.
(Note that there is also a link to API documentation at the top of the GitHub page of cuIBM, next to the title.)
I have also added community guidelines in the file `CONTRIBUTING.md`.

Those changes are now available in the branch `joss-reviews-hotfix-0.1.2`.
This branch will be merged into `master` and a new release (v0.1.2) will be created once we have satisfied all your requests.

**Comment by kyleniemeyer** [comment]
Thanks for your reviews @arghdos and @OSUmageed! Do the documentation and community guideline changes that @mesnardo made satisfy those concerns? If so, you can close those issues / check the boxes above.

@mesnardo It seems like the performance claim is the only remaining issue—do you have a response on that?

**Comment by arghdos** [comment]
@kyleniemeyer I've closed my issues and updated the review

**Comment by OSUmageed** [comment]
@kyleniemeyer @arghdos @mesnardo I closed my issue as well.

**Comment by mesnardo** [comment]
@arghdos @OSUmageed Thank you for your reviews and closing the issues.

@kyleniemeyer Concerning the performance and the fact that the CPU routines are not functional, I am waiting for the input from @anushkrish who has run the tests to compare the GPU version of cuIBM to the CPU one.

**Comment by kyleniemeyer** [comment]
@mesnardo ok great! That looks like the only outstanding issue.

**Comment by arfon** [comment]
@kyleniemeyer - is this good to accept now?

**Comment by kyleniemeyer** [comment]
@arfon no, still waiting on this outstanding performance issue to be resolved

**Comment by labarba** [comment]
@kyleniemeyer @arghdos There is no outstanding performance issue — the CPU version is not maintained, not supported, and there is a note to that effect in the README. The reviewer is referencing an *old* preprint from many years ago, not *this* paper. We make no CPU performance claims in this paper, nor on the repository of the software. The software repository _only_ includes performance claims in the form of runtime for the examples, on GPU.

**Comment by kyleniemeyer** [comment]
OK, @labarba @mesnardo, as long as there is no performance claim vs CPU. So, it looks like the software itself is good to go.

However, the [article itself](https://github.com/barbagroup/cuIBM/blob/master/paper.md) has a few outstanding issues: 

- [ ] The references in the paper are missing DOIs (where appropriate). Could you add those to the `paper.bib` file?
- [ ] Could you add a more general statement of need perhaps to the beginning of the article, that a broader audience would understand? For example, something about fluid-structure interaction would be appropriate, which isn't currently mentioned.
- [ ] The sentence beginning with "Among them" is missing a verb.

Once those things are resolved, this submission will be ready for publication.

**Comment by labarba** [comment]
We'll look into that, and we also just discussed with @mesnardo that we'll remove the CPU routines from the master branch and leave it in a `deprecated` branch.

BTW, this software does _not_ compute (two-way) fluid-structure interaction (only prescribed body motions).

**Comment by labarba** [comment]
(And, as you heard me rant at SciPy, we really have to _stop_ talking about GPU vs. CPU comparisons. This is a fixation that is an unfortunate consequence of the vendor marketing, and it no longer makes sense.)

**Comment by kyleniemeyer** [comment]
@labarba sounds good, thanks! And yes, I am well aware of this—hence the development of our fork of cuIBM that can do so 😀. Perhaps "fluid-structure interaction" would be misleading then, but a statement about simulating flows interacting with moving bodies (e.g.) would be helpful.

**Comment by mesnardo** [comment]
@kyleniemeyer , we have removed the CPU routines, there are now in a separate branch called `deprecated-cpu-routines`.
Also, we have updated the paper and the references with the suggestions you made above.

**Comment by kyleniemeyer** [comment]
OK, looks good! One last thing: the Known Issues at the bottom still says "CPU routines do not work." Could you correct that, and then provide a DOI for the archived software here?

**Comment by mesnardo** [comment]
You should be looking at the branch `joss-reviews-hotfix-0.1.2`; it includes all the changes made during the peer-review.
We are waiting for acceptance before making the new archive and getting a DOI.

**Comment by kyleniemeyer** [comment]
@mesnardo ah, got it—I was indeed looking at the master README. 

OK, this is accepted for publication, so please merge that branch and archive the software.

**Comment by mesnardo** [comment]
@kyleniemeyer Branch `joss-reviews-hotfix-0.1.2` has been merged into `master`, cuIBM-0.1.2 has been released and uploaded to [Zenodo](https://doi.org/10.5281/zenodo.832338) (10.5281/zenodo.832338).

**Comment by kyleniemeyer** [comment]
@whedon commands

**Comment by whedon** [comment]
Here are some things you can ask me to do:

```
# List all of Whedon's capabilities
@whedon commands

# Assign a GitHub user as the reviewer of this submission
@whedon assign @username as reviewer

# List the GitHub usernames of the JOSS editors
@whedon list editors

# List of JOSS reviewers together with programming language preferences and domain expertise
@whedon list reviewers

# Change editorial assignment
@whedon assign @username as editor

# Set the software archive DOI at the top of the issue e.g.
@whedon set 10.0000/zenodo.00000 as archive

# Open the review issue
@whedon start review
```

:construction: **Important** :construction:

This is all quite new. Please make sure you check the top of the issue after running a @whedon command (you might also need to refresh the page to see the issue update).

**Comment by kyleniemeyer** [comment]
@whedon set 10.5281/zenodo.832338 as archive

**Comment by whedon** [comment]
OK. <a href="http://dx.doi.org/10.5281/zenodo.832338" target="_blank">10.5281/zenodo.832338</a> is the archive.

**Comment by kyleniemeyer** [comment]
Hi @arfon this is ready to publish!

**Comment by arfon** [comment]
@arghdos many thanks for your review here and @kyleniemeyer for editing this submission ✨ 

@mesnardo - your paper is now accepted into JOSS and your DOI is http://dx.doi.org/10.21105/joss.00301 ⚡️ 🚀 💥

**Comment by kyleniemeyer** [comment]
Thanks @arghdos and @OSUmageed!

**Comment by mesnardo** [comment]
Yay!
Thank you very much @kyleniemeyer , @arghdos , and @OSUmageed for your review of cuIBM and your useful suggestions!
